text_generation/causal_lm/cpp/README.md

```diff
@@ -4,6 +4,26 @@ These examples showcase inference of text-generation Large Language Models (LLMs
 
 ## How it works
 
+### Stateful LLM
+
+The examples infer a stateful model. The stateful model keeps internal state which is updated after every inference. That state is used as additional implicit model inputs. It is opposed to stateless model which requires passing state tensors explicitly. A more detailed introduction to stateful models can be found in https://github.com/openvinotoolkit/openvino/pull/22410. TODO: update link
+
+The intermediate inference results are usually stored in KV (key value) cache in case of LLMs. This cache is represented by the corresponding inputs and outputs. Stateless model approach requires manipulating these outputs to use them as inputs for the next inference. Stateful models remove that burden and keep KV cache internally. Although the cache is still accessible with state API.
+
+Hiding KV cache introduces a peculiarity for beam search algorithm. Beam search suggests batched inference on multiple beams. The design described here so far would result in generating multiple independent sequences of tokens. Beam search algorithm, on the other hand, requires removing some of the ongoing beams and splitting other beams to multiple branches. Beam removal requires deleting corresponding KV cache entry and beam splitting requires copying corresponding KV cache values.
+
+To provide a possibility to implement beam search without accessing model internal state, a stateful LLM converted with `optimum-intel` or [llm_bench](../../../llm_bench/python/) introduces additional 1-dimentional `beam_idx` input. `beam_idx` must contain indices of elements in a batch which are supposed to be selected and evolve during next beam search iteration. Suppose there are two running beams. To proceed generating both beams at the next iteration, `beam_idx` values must be `[0, 1]`, pointing to batch elements `0` and `1`. To drop the last beam and split the other beam in two, `beam_idx` must be set to `[0, 0]`, this results in utilizing only the part of KV cache corresponding to zeroth element in the batch. The process of selecting appropriate entries in cache is called Cache Reorder.
```

Comment:
Provide a picture that illustrates how beam_idx fork sequence and removes sequence. Do it in a separate PR please.